### ISSUE TITLE
[Bug]: PC0030 icw RecordRef

### ISSUE THREAD

**Issue by guidorobben** [ALCops/Analyzers#497]
### Analyzer

PlatformCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0030

### Description

When using a recordref on multiple tables, the warning of PC0030 is not needed in my opinion. The data comes from different tables and different fields. That is the reason RecordRefs are used. You dont know what data you are getting. In my opinion the SetLoadFields doesnt make sence. 

<img width="1278" height="555" alt="Image" src="https://github.com/user-attachments/assets/328c4eed-3edb-4287-ad7d-7a3ab2d82de7" />

### AL code to reproduce

```al
    procedure GetDocumentStatus(DocumentRecordID: RecordId) DocumentStatus: Enum "Purchase Document Status"
    var
        DocumentRecordRef: RecordRef;
    begin
        DocumentStatus := DocumentStatus::Open;

        DocumentRecordRef := DocumentRecordID.GetRecord();
        if not DocumentRecordRef.FindFirst() then
            exit;

        case DocumentRecordRef.Number() of
            Database::"Purchase Header":
                exit(DocumentRecordRef.Field('Status').Value());
            Database::"Purch. Inv. Header",
            Database::"Purch. Cr. Memo Hdr.":
                exit(DocumentRecordRef.Field('Status Approval').Value());
        end;
    end;

```

**Comment by Arthurvdv** [comment]
Thanks for reporting this! After looking into it, I don't believe this is a false positive — `SetLoadFields` can be applied here.

The table number is already known after `GetRecord()`, so you can check `Number()` before the read and set the appropriate load fields per table:

```al
DocumentRecordRef := DocumentRecordID.GetRecord();

case DocumentRecordRef.Number() of
    Database::"Purchase Header":
        DocumentRecordRef.SetLoadFields(DocumentRecordRef.Field('Status').Number);
    Database::"Purch. Inv. Header",
    Database::"Purch. Cr. Memo Hdr.":
        DocumentRecordRef.SetLoadFields(DocumentRecordRef.Field('StatusApproval').Number);
end;

if not DocumentRecordRef.FindFirst() then
    exit;
```

**Comment by guidorobben** [comment]
I agree with you that it is possible. But it will look like this. hmmm. Doesn't look that pretty. with a lot of code overhead. 

```al
 procedure GetDocumentStatus(DocumentRecordID: RecordId) DocumentStatus: Enum "Purchase Document Status"
  var
      DocumentRecordRef: RecordRef;
  begin
      DocumentStatus := DocumentStatus::Open;

      DocumentRecordRef := DocumentRecordID.GetRecord();
      case DocumentRecordRef.Number() of
          Database::"Purchase Header":
              DocumentRecordRef.SetLoadFields(DocumentRecordRef.Field(Status).Number);
          Database::"Purch. Inv. Header",
          Database::"Purch. Cr. Memo Hdr.":
              DocumentRecordRef.SetLoadFields(DocumentRecordRef.Field(StatusApproval).Number);
      end;

      if not DocumentRecordRef.FindFirst() then
          exit;

      case DocumentRecordRef.Number() of
          Database::"Purchase Header":
              exit(DocumentRecordRef.Field('Status').Value());
          Database::"Purch. Inv. Header",
          Database::"Purch. Cr. Memo Hdr.":
              exit(DocumentRecordRef.Field('Status Approval').Value());
      end;
  end;
```

**Comment by Arthurvdv** [comment]
The only thing that seems to differs is the field name, so we can pull this into a variable and collapse to one case I think;

```al
procedure GetDocumentStatus(DocumentRecordID: RecordId) DocumentStatus: Enum "Purchase Document Status"
var
    DocumentRecordRef: RecordRef;
    StatusFieldNo: Integer;
begin
    DocumentStatus := DocumentStatus::Open;

    DocumentRecordRef := DocumentRecordID.GetRecord();
    case DocumentRecordRef.Number() of
        Database::"Purchase Header":
            StatusFieldNo := DocumentRecordRef.Field('Status').Number;
        Database::"Purch. Inv. Header",
        Database::"Purch. Cr. Memo Hdr.":
            StatusFieldNo := DocumentRecordRef.Field('Status Approval').Number;
        else
            exit;
    end;

    DocumentRecordRef.SetLoadFields(StatusFieldNo);
    if not DocumentRecordRef.FindFirst() then
        exit;

    DocumentStatus := DocumentRecordRef.Field(StatusFieldNo).Value();
end;
```

**Comment by guidorobben** [comment]
I agree with you. This is a good solution. Never knew you could use numbers in the setloadfields. So it's not a false positive. Case closed
